lib/api/objectCopy.js

```diff
@@ -0,0 +1,288 @@
+                        sourceObjMD.location : [{ key: sourceObjMD.location }];
+                    }
+
+                    if (sourceObjMD['x-amz-server-side-encryption']) {
+                        for (let i = 0; i < dataLocator.length; i++) {
+                            dataLocator[i].masterKeyId = sourceObjMD
+                            ['x-amz-server-side-encryption-aws-kms-key-id'];
+                            dataLocator[i].algorithm =
+                                sourceObjMD['x-amz-server-side-encryption'];
+                        }
+                    }
+                    return next(null, storeMetadataParams, dataLocator);
+                });
+        },
+        function checkDestAuth(storeMetadataParams, dataLocator, next) {
+            return services.metadataValidateAuthorization(valPutParams,
+                (err, destBucketMD, destObjMD) => {
+                    if (err) {
+                        log.debug('error validating put part of request',
+                        { error: err });
+                        return next(err);
+                    }
+                    const flag = destBucketMD.hasDeletedFlag()
+                        || destBucketMD.hasTransientFlag();
+                    if (flag) {
+                        log.trace('deleted flag or transient flag ' +
+                        'on destination bucket', { flag });
+                        return next(errors.NoSuchBucket);
+                    }
+                    return next(null, storeMetadataParams,
+                        dataLocator, destBucketMD, destObjMD);
+                });
+        },
+        function getCipherBundle(storeMetadataParams, dataLocator, destBucketMD,
+            destObjMD, next) {
+            const serverSideEncryption = destBucketMD.getServerSideEncryption();
+            if (!serverSideEncryption) {
+                return next(null, storeMetadataParams, dataLocator,
+                    destObjMD, null);
+            }
+            return kms.createCipherBundle(
+                serverSideEncryption,
+                log, (err, cipherBundle) => {
+                    if (err) {
+                        log.debug('error getting cipherBundle');
+                        return next(errors.InternalError);
+                    }
+                    return next(null, storeMetadataParams, dataLocator,
+                        destObjMD, cipherBundle);
+                });
+        },
+        function goGetData(storeMetadataParams, dataLocator,
+            destObjMD, cipherBundle, next) {
+            // skip if source and dest the same or 0 byte object
+            if (sourceIsDestination || dataLocator.length === 0) {
+                return next(null, storeMetadataParams, dataLocator, destObjMD,
+                    cipherBundle);
+            }
+             // dataLocator is an array.  need to get and put all parts
+             // For now, copy 1 part at a time. Could increase the second
```

Comment:
I think, the copy is concurrent. What is made serially here is the connection of streams ends  with ```pipe()``` methods. Once streams are connected they are ready to be consumed by the node event loop